udf.go

```diff
@@ -2,59 +2,39 @@ package udf
 
 import (
 	"errors"
+	"fmt"
 	"io"
+	"unsafe"
 )
 
+// Udf is a wrapper around an .iso file that allows reading its ISO-13346 "UDF" data
 type Udf struct {
-	r        io.ReaderAt
-	isInited bool
-	pvd      *PrimaryVolumeDescriptor
-	pd       map[uint16]*PartitionDescriptor
-	lvd      *LogicalVolumeDescriptor
-	fsd      *FileSetDescriptor
-	root_fe  FileEntryInterface
+	r           io.ReaderAt
+	isInited    bool
+	pvd         *PrimaryVolumeDescriptor
+	pd          map[uint16]*PartitionDescriptor
+	lvd         *LogicalVolumeDescriptor
+	fsd         *FileSetDescriptor
+	root_fe     FileEntryInterface
 	SECTOR_SIZE uint64
 }
 
-func (udf *Udf) PhysicalPartitionStart(partition uint16) (physical uint64) {
-	if udf.pd == nil {
-		panic(udf)
-	} else {
-		return uint64(udf.pd[partition].PartitionStartingLocation)
-	}
-}
-
-func (udf *Udf) LogicalPartitionStart(partition uint16) (logical uint64) {
-	if udf.lvd == nil || len(udf.lvd.PartitionMaps) < 1 {
-		panic(udf)
-	} else {
-		return uint64(udf.lvd.PartitionMaps[partition].PartitionStart)
-	}
-}
-
-func (udf *Udf) GetReader() io.ReaderAt {
-	return udf.r
-}
-
-func (udf *Udf) ReadSectors(sectorNumber uint64, sectorsCount uint64) []byte {
-	buf := make([]byte, udf.SECTOR_SIZE*sectorsCount)
-	read, err := udf.r.ReadAt(buf[:], int64(udf.SECTOR_SIZE*sectorNumber))
-	if err != nil {
-		panic(err)
+// NewUdfFromReader returns an Udf reader reading from a given file
+func NewUdfFromReader(r io.ReaderAt) (*Udf, error) {
+	udf := &Udf{
+		r:        r,
+		isInited: false,
+		pd:       make(map[uint16]*PartitionDescriptor),
 	}
-	/*if readed != int(udf.SECTOR_SIZE*sectorsCount) {
-		panic(readed)
-	}*/
-	return buf[:read]
-}
 
-func (udf *Udf) ReadSector(sectorNumber uint64) []byte {
-	return udf.ReadSectors(sectorNumber, 1)
+	err := udf.init()
+	return udf, err
 }
 
-func (udf *Udf) init() (err error) {
+func (udf *Udf) init() error {
+	var err error
```

Comment:
This change seems to only require extra code down the line

Reply:
Indeed